File: problem/Walking_Robot_Simulation/robotSim.py
class Solution:
    def robotSim(self, commands, obstacles):
        # Store obstacles
        blocked = set()
        for o in obstacles:
            blocked.add((o[0], o[1]))

        # Directions: North, East, South, West
        directions = [
            (0, 1), (1, 0), (0, -1), (-1, 0)
        ]

        x, y = 0, 0
        dir = 0  # initially facing North
        maxDist = 0

        for cmd in commands:
            if cmd == -1:
                dir = (dir + 1) % 4  # turn right
            elif cmd == -2:
                dir = (dir + 3) % 4  # turn left
            else:
                while cmd > 0:
                    nx = x + directions[dir][0]
                    ny = y + directions[dir][1]

                    # check obstacle
                    if (nx, ny) in blocked:
                        break

                    x = nx
                    y = ny

                    maxDist = max(maxDist, x * x + y * y)
                    cmd -= 1

        return maxDist

File: problem/Walking_Robot_Simulation/robotSim.js
/**
 * @param {number[]} commands
 * @param {number[][]} obstacles
 * @return {number}
 */
var robotSim = function(commands, obstacles) {
    const blocked = new Set();
    for (const [ox, oy] of obstacles) {
        blocked.add(`${ox},${oy}`);
    }

    const dx = [0, 1, 0, -1];
    const dy = [1, 0, -1, 0];

    let x = 0, y = 0, dir = 0, maxDist = 0;

    for (const cmd of commands) {
        if (cmd === -1) {
            dir = (dir + 1) % 4;
        } else if (cmd === -2) {
            dir = (dir + 3) % 4;
        } else {
            for (let i = 0; i < cmd; i++) {
                const nx = x + dx[dir];
                const ny = y + dy[dir];

                if (blocked.has(`${nx},${ny}`)) break;

                x = nx;
                y = ny;
                maxDist = Math.max(maxDist, x * x + y * y);
            }
        }
    }

    return maxDist;
};